game-frontend/src/components/user-script/monaco-helper.ts

```diff
@@ -0,0 +1,52 @@
+import { useState, useEffect, useRef } from 'react';
+import { monaco, EditorDidMount } from '@monaco-editor/react';
+import monacoEditor from 'monaco-editor/esm/vs/editor/editor.api';
+
+export type MonacoInstance = typeof monacoEditor;
+type EditorRef = React.MutableRefObject<
+  monacoEditor.editor.IStandaloneCodeEditor | undefined
+>;
+export const useHandleEditorRef = (): [EditorRef, EditorDidMount] => {
+  const editorRef = useRef<monacoEditor.editor.IStandaloneCodeEditor>();
+  const didMount: EditorDidMount = (_, editor) => {
+    editorRef.current = editor;
+  };
+  return [editorRef, didMount];
+};
+
+export const useMonacoInstance = (): MonacoInstance | undefined => {
+  const [monacoInstance, setMonacoInstance] = useState<MonacoInstance>();
+  useEffect(() => {
+    monaco.init().then(m => {
+      setMonacoInstance(m);
+    });
+  }, []);
+  return monacoInstance;
+};
+
+export const useSaveHotKey = (
+  editorRef: EditorRef,
+  save: (value: string) => void
+): void => {
+  const monacoInstance = useMonacoInstance();
+  const SAVE_ACTION_ID = 'DARWIN_SAVE';
+
+  useEffect(() => {
+    if (editorRef && editorRef.current && monacoInstance) {
+      const editorSaveAction = editorRef.current.getAction(SAVE_ACTION_ID);
+      if (!editorSaveAction || editorSaveAction.run !== save) {
```

Comment:
What's the idea behind the `editorSaveAction.run !== save`?

Reply:
To change the editorsAction when the save method changes

Reply:
And how exactly does this work when you assign a new function to the run property of the action? (See line 45)
```
run: editor => {
  save(editor.getValue());
},
```

And therefore `editorSaveAction.run` will **never** be `save`. (If you'd have assigned `save` to `run:`, yeah that would work - and that's probably what you meant to do?)
Or am I missing something?

Reply:
Yeah, you're right, that's what i meant to do